Improvements to Gitpod workspace and addition of Makefile for automation

ThomasSanson opened the pull request:
Close #2734

This pull request encompasses a series of enhancements designed to streamline project workflow and improve the overall Gitpod workspace experience. The following summarizes the key changes:

1. **Addition of a Makefile**: A Makefile was added to the project to automate certain tasks, providing a more efficient workflow. [Commit Link](https://github.com/oxsecurity/megalinter/commit/4bd12833962f1a8dacff062dad3f9c930dbb9f46)

2. **Global configuration for the Makefile**: A global configuration was added to the Makefile to ensure consistency and reduce potential errors. [Commit Link](https://github.com/oxsecurity/megalinter/commit/8153b1c7418130f68de26d39bd37b57390f7fa07)

3. **Help command added to Makefile**: To improve usability, a help command was introduced to the Makefile. [Commit Link](https://github.com/oxsecurity/megalinter/commit/14ede58bd220898cbfee1758df9b7300dc2f3ec1)

4. **Addition of .venv to .gitignore**: The .venv directory was added to the .gitignore file to prevent tracking of changes in this folder, further enhancing the project's maintainability. [Commit Link](https://github.com/oxsecurity/megalinter/commit/4f88eaa7219f04ef7c5091fafe5c3b55eb370bd4)

5. **Addition of a Gitpod configuration file**: A Gitpod configuration file was added to facilitate the use of Gitpod with this project, streamlining the setup process for new contributors. [Commit Link](https://github.com/oxsecurity/megalinter/commit/f4f4908ee5b72795fcbbc51f697ff35009e8570a)

Further additions include a Dockerfile for the Gitpod workspace, targets for building, linting, and testing in Gitpod, a Python virtual environment, and updated contribution guidelines.

In total, there are 25 commits affecting 18 files.

ThomasSanson commented:
Hello @nvuillam,

I'm having trouble with this pull request and I'm unable to resolve the error. Could you please take a look and help me understand what's going wrong with Trivy ?

Thanks in advance

nvuillam commented:
@ThomasSanson you install new tools that have dependencies identified as flawed by Trivy

Most of the time this is a false positive or not impacting for use in MegaLinter, but each issue must be observed individually to make sure of that :) You can also try to upgrade versions of impacted dependencies

ThomasSanson commented:
@nvuillam, alright, but I'm struggling to understand where the issue lies with the new tools.

Trivy is indicating that there is only one error, but I can't pinpoint which one it is.

Could you possibly guide me with a link to the line that's causing trouble?

Starting from this page:
https://github.com/oxsecurity/megalinter/actions/runs/5231399932/jobs/9445518008?pr=2737#step:4:2584

Because in this log, I only see files located in ".automation/tests".

The only file that is in this log and that I've moved is "requirements.dev.txt", which I moved to ".config/python/dev". But that's just a relocation, I haven't added anything...
Moreover, if I upgrade "pymdown-extensions" to version 10.0, the error still persists...

I've put in the effort to try and fix it, but honestly, I'm lost with this log output.

nvuillam commented on .mega-linter.yml:
```diff
@@ -25,22 +25,22 @@ DISABLE_ERRORS_LINTERS:
 REPOSITORY_TRIVY_ARGUMENTS:
   - "--skip-dirs"
-  - ".automation/test"
+  - ".automation/test .venv"
```
@ThomasSanson here you forgot to go in a new line, so you messed with trivy arguments, that's what explains all these errors :)

nvuillam replied:
It also seems that --skip-dirs must be written for each dir
https://aquasecurity.github.io/trivy/v0.42/docs/configuration/skipping/#skip-directories

ThomasSanson replied:
Ah okay, I just understood. I was so focused on the single error from Trivy that I ended up getting confused in the process. In fact, I was under the impression that it was normal to detect errors in the .automation/test directory, like for example:

```bash
  .automation/test/docker/bad/Dockerfile (dockerfile)
  ===================================================
  Tests: 20 (SUCCESSES: 17, FAILURES: 3, EXCEPTIONS: 0)
  Failures: 3 (UNKNOWN: 0, LOW: 1, MEDIUM: 0, HIGH: 0, CRITICAL: 2)
```

Thank you for your help ! And also for the solution !

ThomasSanson replied:
Done

nvuillam commented on .config/python_version.config:
```diff
@@ -0,0 +1 @@
+PYTHON_VERSION=3.10
```
We're currently in 3.11.4 :)

ThomasSanson replied:
done here 
https://github.com/oxsecurity/megalinter/pull/2737/commits/e112cb0146277d59410e3261572e087bfbd09251

nvuillam commented on .github/workflows/gitpod.yml:
```diff
@@ -0,0 +1,36 @@
+on:
+  push:
+    branches:
+      - master
```
We use main branch, not master ^^

ThomasSanson replied:
my bad, fix here :
https://github.com/oxsecurity/megalinter/pull/2737/commits/e4fb2053eaf896aebf54e33a67ce6eabce369fbb

nvuillam commented on .github/workflows/gitpod.yml:
```diff
@@ -0,0 +1,36 @@
+      - master
+  pull_request:
+    branches:
+      - master
```
We use main branch, not master ^^

ThomasSanson replied:
my bad, fix here :
https://github.com/oxsecurity/megalinter/pull/2737/commits/e4fb2053eaf896aebf54e33a67ce6eabce369fbb

nvuillam commented on .mega-linter.yml:
```diff
@@ -25,22 +25,24 @@ DISABLE_ERRORS_LINTERS:
 MARKDOWN_FILTER_REGEX_EXCLUDE: '(license\.md|docs/licenses)'
 MARKDOWN_MARKDOWN_LINK_CHECK_DISABLE_ERRORS: true
-SPELL_MISSPELL_FILTER_REGEX_EXCLUDE: '(\.automation/generated|docs/descriptors|used-by-stats|docs/used-by-stats\.md|.github/linters/valestyles)'
+SPELL_MISSPELL_FILTER_REGEX_EXCLUDE: '(\.automation/generated|\.venv|docs/descriptors|used-by-stats|docs/used-by-stats\.md|.github/linters/valestyles)'
```
MISSPELL has been removed from MegaLinter v7, plz can you remove this config line ?

ThomasSanson replied:
done here : https://github.com/oxsecurity/megalinter/pull/2737/commits/3532da01529b0907a148f32951c7b8c1e2081d71

nvuillam reviewed (changes requested):
Plz check comments :)

+ Please update changelog.md

As it is an impacting PR, @echoix @Kurt-von-Laven @bdovaz I'd love to have your opinion / validation :)

echoix commented:
Did you ever hear about the .python-version files in the root directory? I know it is used by pyenv, and also for the setup Python action in GitHub (that would help the build-command.yml action to keep getting updated)

echoix commented on .config/gitpod/Dockerfile:
What's the benefit of a custom docker image for gitpod, instead of basing it from one of their generic images (that they already have cached, set up and maintained, that contains all workaround for the product). I'm not even considering the build time on their side to start. 

I know that for GitHub Codespaces, devcontainers not based from their generic image count against storage, and are slower to load (or don't load at all if at one point the build file is no longer valid)

ThomasSanson replied:
Absolutely, @echoix.

Here are some strong arguments for using a custom Docker image with Gitpod:

1. **Precise development environments**: Custom Docker images allow you to define very specific and precise development environments. You can include project-specific dependencies, development tools, or configurations that may not be part of Gitpod's generic images.

2. **Reproducibility**: By using your own Docker image, you ensure that the development environment is reproduced identically each time. This can help avoid the "it works on my machine" issues that can arise when developers work on slightly different environments.

3. **Control**: Having your own Docker image gives you more control over your development environment. You can choose when and how to update dependencies, which can be important for compatibility or security reasons.

4. **Flexibility**: Custom Docker images are highly flexible. You can tailor them to meet very specific needs not covered by Gitpod's generic images.

5. **Optimised start-up time**: You can build and push your custom Docker image to a Docker image registry, whether public or private. Then, you can configure Gitpod to use this pre-built image. This means Gitpod doesn't need to build the image from scratch each time a new workspace is started, which can significantly reduce start-up time.

6. **Image version management**: When you push your Docker image to a registry, you can also use tags to manage different versions of your image. This can be useful if you need to maintain multiple versions of your development environment for different branches or versions of your application.

While using a custom Docker image can indeed lead to longer load times and consume more storage than using a Gitpod generic image, it's a trade-off between flexibility and control on one hand, and convenience and speed on the other. 

By storing and uploading the image, you can combine the best of both worlds and create a development environment that precisely meets your needs while minimising start-up time and storage usage.

echoix replied:
Thanks again for your greatly descriptive replies, I'd need a little time to correctly read out and address that new info 😃

echoix commented on Makefile:
I wasn't expecting a make-based solution at all. Your makefiles are really neat, and seem like they come from a real master, being so clean and all. 

But I'm just hesitant that we would be able to keep maintaining them correctly in the future, as I don't know for others, but it's not really in our skill set, and I think some of us are maintaining based on Windows.

I know it might be kinda late for it, and that it might not be as tool-universal that's a make-based solution could work on Linux-based environnements, but what about having some setup on the IDE-side, so that a VSCode setup in the repo, would work:
- locally (VsCode, VsCodium, etc, but also on Windows like what people not using web tools might use)
- VsCode on the web -> codespaces, I'm not sure that one of our tools might be available as extensions running only on web (in the browser, without the server back end)
- Gitpod (on the web, it is still a VsCode, but can be opened locally in vscode, or even PyCharm/a jetbrains IDE)

I'm thinking about defining tasks of the sort in that config.


I also understand a point of view that for new contributors, it might be easier to propose just have one way of doing things, that's simpler to understand/support. However, I think it's easier to understand if it's something that is more common across projects (than a really custom solution) and might be closer to what they use already, if they don't really read the docs before/it's too long.

I think it's great to use web services like Gitpod or Codespaces, but not on the point of relying 100% on them, ie, being at their mercy if anything changes. That might be also said against VSCode, but at the scale of how many people use it all the time, I'm not scared that it will break or go away real fast.

ThomasSanson replied:
@echoix,

Thank you for your message concerning my Makefile. I appreciate your thoughtful feedback and the time you've taken to articulate your perspective.

I wanted to clarify a few points in response to your comments. While I appreciate the elegance of Makefile, I would actually prefer to transition to Taskfile (https://taskfile.dev/), which is a modern tool specifically built for managing tasks. It's written in Go, which allows for broader functionality and cleaner syntax. However, I haven't yet had the opportunity to fully investigate this alternative.

Furthermore, I wanted to underline that Makefiles are in no way tied to Gitpod or VSCode. For me, they serve as a universal entry point that provides each contributor with the freedom to proceed as they see fit. The idea is to utilise Makefile as the 'source of truth', and I would even go as far as suggesting that it should be what we run from our CI pipeline.

Regarding the development across different platforms (Windows, Linux, web development, etc.), I understand your concerns. In my view, there needs to be a clear and simple development process for everyone, while allowing each individual to handle their OS-specific features (be it Zsh, Windows, etc.). In this context, a Makefile isn't more restrictive than a "build.sh".

As for the maintainability of Makefiles, I have strived to keep them as simple as possible. I anticipate no significant issues or complexity in terms of maintenance. We should, however, remain open to revisiting this should any unexpected difficulties arise.

Thank you once again for sharing your thoughts.

ThomasSanson replied:
The `.config` directory is emerging as a standard across various projects, especially in light of the multitude of root-level files that one often finds in such undertakings. I find this concept highly favourable, hence my enthusiasm in advocating for it.

I wholeheartedly agree with your point about adhering to established standards. In line with this, plans are underway to modify the `.python-version` file to reside at the root level of the repository. 

Regarding the specific case of the `.config/python/dev/requirements.txt` file, the primary motivation behind this arrangement is to accommodate Renovate. By default, Renovate searches for all `**/requirements.txt` files, rather than `requirements.dev.txt` files. Hence, this structure.

echoix replied:
> In this context, a Makefile isn't more restrictive than a "build.sh".

I personally don't really hang on to that build.sh (in fr-CA: je n'y tiens pas particulièrement). I'm not sure about it, but it seems like a remanent of Super-linter, where we come from. It doesn't do very much

echoix replied:
I've rethought about it and my concerns of integrating more into VS code config doesn't exclude using makefiles. We could just set up calls to à make subcommand inside these. For another time though.

echoix commented on .config/python/dev/requirements.txt:
I understand that with the organization of the pr, having that file there is the correct place to have everything organized. 

But I think the convention or common usage of Python requirement files is to be at the root of the repo. We don't have one, (since our repo isn't completely a Python project, if we would have one, it would be in the megalinter subfolder, the root of the Python project). Sometimes we see a separate file for other dependencies. That's what we find. And it's quite common across various instructions on how to use a dependencies file, even if it's not the plain one, but for dev dependencies.

In summary, I fear that we are specializing us too much from mainstream patterns. Correct me if I didn't see enough repo organizations in either Python-based projects or Ci-based projects.

ThomasSanson replied:
The `.config` directory is emerging as a standard across various projects, especially in light of the multitude of root-level files that one often finds in such undertakings. I find this concept highly favourable, hence my enthusiasm in advocating for it.

I wholeheartedly agree with your point about adhering to established standards. In line with this, plans are underway to modify the `.python-version` file to reside at the root level of the repository. 

Regarding the specific case of the `.config/python/dev/requirements.txt` file, the primary motivation behind this arrangement is to accommodate Renovate. By default, Renovate searches for all `**/requirements.txt` files, rather than `requirements.dev.txt` files. Hence, this structure.

echoix replied:
> Regarding the specific case of the .config/python/dev/requirements.txt file, the primary motivation behind this arrangement is to accommodate Renovate. By default, Renovate searches for all **/requirements.txt files, rather than requirements.dev.txt files. Hence, this structure.

The possibility to have a dependency updater check that file would be great indeed! Its already plan to there eventually. But for this, it might cause some problems in the short/medium term; We could have dependencies updated newer than what our project supports. I would need to verify if the daily update task, or the one that updates the docs loads this file, and if yes, then there's no problem.

For Python dependencies inside the image, that's one medium/long term goal to have a "bidirectional"-ish way to have our dependencies defined inside our descriptors (for the flavors and all extra metadata), but also outputted in a file readable by dependabot/renovate or other, and a way to easily handle updates to the updated files (by dependabot) back to our descriptor files. It's not figured out yet. And it would allow to pin the versions used at a point in time inside the contents of a repo, and be able to rebuild an image with the same dependencies available at that time.

echoix replied:
I think your implementation was well founded.

echoix commented:
By the way, your way of committing is really clean and professional. I think it's the conventional commits way, used in Angular am I right?

ThomasSanson commented:
@echoix, thank you for bringing up the `.python-version` file. 
I wasn't aware of this specific feature of pyenv. Adopting this practice for the project seems like an excellent idea, as it could simplify setting up the Python environment for every contributor. 

I will endeavour to submit a modification to incorporate this improvement into the code over the coming week. 

I'll keep you updated once it has been done.

Could you please annotate the `.config/python_version.config` file to facilitate our discussion about it ?

Thanks again for your invaluable suggestion.

echoix commented on .config/python_version.config:
See https://github.com/oxsecurity/megalinter/pull/2737#issuecomment-1586267956

ThomasSanson replied:
Done here :
https://github.com/oxsecurity/megalinter/pull/2737/commits/b84a7d54fc1850a03914916085013ea60c1347d6
https://github.com/oxsecurity/megalinter/pull/2737/commits/4cbc9d7c7933270a02daf1244b0f6e3cfda70d22
https://github.com/oxsecurity/megalinter/pull/2737/commits/870b4d8163fdc8fde2874d92a565e52a01972a94
https://github.com/oxsecurity/megalinter/pull/2737/commits/c67b14697d90e1ca7e33eaa513daf060e8246400

ThomasSanson replied:
Hello @echoix,

I've looked into pyenv more thoroughly and adjusted the .python-version file accordingly.

However, a question has been nagging me: wouldn't it be better to choose an installation via pyenv over a direct Python installation in the Dockerfile?

And of course, the corresponding documentation would need to be updated.

So instead of this:
```bash
...
    python3-pip \
    "python${PYTHON_VERSION}" \
    "python${PYTHON_VERSION}-venv" \
...
```

We could have something like this:
```bash
...
# Install pyenv
RUN set -ex \
    && curl https://pyenv.run | bash \
    && pyenv update \
    && pyenv install "${PYTHON_VERSION}" \
    && pyenv global "${PYTHON_VERSION}" \
    && pyenv rehash
```

What do you think?

echoix replied:
> However, a question has been nagging me: wouldn't it be better to choose an installation via pyenv over a direct Python installation in the Dockerfile?
> 
> And of course, the corresponding documentation would need to be updated.
> 
> So instead of this:
> 
> ```shell
> ...
>     python3-pip \
>     "python${PYTHON_VERSION}" \
>     "python${PYTHON_VERSION}-venv" \
> ...
> ```
> 
> We could have something like this:
> 
> ```shell
> ...
> # Install pyenv
> RUN set -ex \
>     && curl https://pyenv.run | bash \
>     && pyenv update \
>     && pyenv install "${PYTHON_VERSION}" \
>     && pyenv global "${PYTHON_VERSION}" \
>     && pyenv rehash
> ```
> 
> What do you think?

I'm not sure if its relevant now, but yeah, for sure, inside gitpod we might as well use pyenv directly. But for our alpine images, not really, were not going the pyenv route... I think that it might be clearer to use one type of enviroment managing only.
I never heard of pyenv rehash, and I'm not too sure how pyenv global works (or the implications of using it). If it is what I think, that it changes/sets up the global python version, we might as well make use of the `pyenv local` instead, so if they reuse these instructions on their local install, they don't end up poisoning their global python installs (that will make problems for their contributions to megalinter if this ends up that pip uses other older dependencies, and that it wouldn't be like other setups)

echoix commented:
> @echoix, thank you for bringing up the `.python-version` file. 
> 
> I wasn't aware of this specific feature of pyenv. Adopting this practice for the project seems like an excellent idea, as it could simplify setting up the Python environment for every contributor. 
> 
> 
> 
> I will endeavour to submit a modification to incorporate this improvement into the code over the coming week. 
> 
> 
> 
> I'll keep you updated once it has been done.
> 
> 
> 
> Could you please annotate the `.config/python_version.config` file to facilitate our discussion about it ?
> 
> 
> 
> Thanks again for your invaluable suggestion.

Also, some tools understand it directly, which is nice

ThomasSanson commented:
> By the way, your way of committing is really clean and professional. I think it's the conventional commits way, used in Angular am I right?

Indeed, you are correct @echoix. 

I do adhere to the Conventional Commits approach (https://www.conventionalcommits.org/en/v1.0.0/) as it provides a disciplined and structured methodology for committing changes. 

I also strive to describe with precision what I've accomplished in each commit. I find it crucial for maintaining clarity and professionalism in our version control system. 

Thank you for noticing and appreciating this aspect of my work !

nvuillam commented on .github/workflows/gitpod.yml:
```diff
@@ -0,0 +1,36 @@
+      - name: Install dependencies
+        run: make bootstrap
+
+      - name: Run Giptod tests
```
Giptod ? :p

nvuillam commented:
I've again reviewed the code and it seems really nice, and I believe in the added value it can bring to MegaLinter.

3 things to solve before considering merging it:
- Is it "added features", it does not replace the existing way to contribute on MegaLinter, right ? ( meaning I can continue contributing the way I do today, for example)
- @ThomasSanson if we merge it, will you stay around in case evolutions are necessary , and internally it seems we don't have the knowledge about Makefiles ?
- @bdovaz @Kurt-von-Laven I'd really like to also have your opinion :)

nvuillam commented on .github/CONTRIBUTING.md:
```diff
@@ -3,49 +3,60 @@
+
+#### 4.2 Desktop
+
+Install [make](https://www.gnu.org/software/make/), [Python3.10](https://www.python.org/), [venv](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/), [docker](https://docs.docker.com/engine/install/ubuntu/) and [nodejs](https://github.com/nodesource/distributions/tree/master).
```
not 3.10 ^^

ThomasSanson commented:
> I've again reviewed the code and it seems really nice, and I believe in the added value it can bring to MegaLinter contribution experience by reading the detailed exchange between @echoix & @ThomasSanson .
> 
> 3 things to solve before considering merging it:
> 
> * Is it "added features", it does not replace the existing way to contribute on MegaLinter, right ? ( meaning I can continue contributing the way I do today, for example)
> * @ThomasSanson if we merge it, will you stay around in case evolutions are necessary , and internally it seems we don't have the knowledge about Makefiles ?
> * @bdovaz @Kurt-von-Laven I'd really like to also have your opinion :)

Hello @nvuillam,

Thank you for taking the time to review the code and for recognising the value it could bring to contributing to MegaLinter.

To answer your questions:

* Indeed, this is an "added feature". The intent is not to replace the existing way of contributing to MegaLinter, but to complement it. You can, of course, continue to contribute in the way you are accustomed to.

* If we decide to merge the code, I commit to being available for any necessary changes. I understand that the internal knowledge about Makefiles may be lacking, and I would be more than happy to provide assistance and advice in this regard.

As for the Makefile, your concern is completely understandable. The team might feel more comfortable with a Taskfile or even npm. However, using npm could introduce additional complexity, especially when it comes to managing Python and the virtual environments necessary for the project.

I would like to clarify that the goal of this proposal is to provide a single source of truth, not to impose a constraint. Thus, a new contributor could simply launch a gitpod (once they've created their account), make their contribution, then run `make tests` or `make tests-fast` to ensure everything is functioning as expected. If someone prefers not to use gitpod, they can still use `make bootstrap` to initialise their environment. Finally, if someone prefers to manage their workspace in their own way (and handle maintenance and dependency issues), that is also an option. In all cases, the pull request pipeline serves as a safeguard. However, it is clear that using gitpod and running `make tests` allows for a swift and effective rollback if needed.

I look forward to hearing the thoughts of @bdovaz and @Kurt-von-Laven on this matter. Thank you once again for your constructive feedback.

echoix commented:
I'm starting to take a look at your branch and opening it in Gitpod. I thought about it during the week. And now I think I'm ok with the makefile. I'll write more soon

bdovaz commented:
In my case, it is the first time I hear about Gitpod but reading you and reading a little about Gitpod, it sounds very good.

Anything that eases the barrier of entry for contributors is welcome.

echoix commented on .config/gitpod/Dockerfile:
```diff
@@ -0,0 +1,100 @@
+    "mega-linter-runner@${MEGA_LINTER_RUNNER_VERSION}" \
+    "markdown-table-formatter@${MARKDOWN_TABLE_FORMATTER_VERSION}"
+
+###########################
+#         cleanup         #
+###########################
+FROM npm-installer as cleanup
+
+RUN apt-get clean && rm -rf /var/lib/apt/lists/* tmp/* 
+
```
This has a conceptual error. (We kinda became Dockerfile specialists with time).
Even though we might think that since it a separate stage, and that we often use that pattern to build optimized images, it doesn't do exactly what you might think.

Since it is a "FROM", then this stage will create a new image by using `cleanup` as a base image, like if it was a published image, but in this case, it is defined in the Dockerfile. Docker images work essentially by applying layers on top of each other, each layer being immutable. Just like if each layer was a tar file, and extracted in the same folder. (It is really like that it works).

So, just before the RUN apt-get clean from line 83, the layers of the image will be all the layers that `cleanup` consists of, and at line 83, we are adding a new layer, that will, once applied, delete files in the filesystem of the image.

That means that what will be downloaded and extracted will contain the files, and later on when finishing pulling, we will remove them. So there are no space gains at all, since they had to be included in the layer like at lines 17-20, 53, and 71, but don't appear in the final state. It is "wasted space". The solution for this is the (ugly) pattern that you often see in Dockerfiles, that you chain the clean commands directly in the same "RUN" as you created them. Each instruction in a Dockerfile makes a layer, that won't change.


Edit: I read a little too fast. Finally you used a `COPY --from` at the last stage that was based off the original one.  So it doesn't really matter inside the final image, since all the stages will be discarded. But I'm not sure that the user added applied on line 91 will really work well if you copy and replace the whole filesystem. If you are to base your image on `ubuntu:jammy`, and then completely replace the whole filesystem with your contents, why need to have the final image download two times the contents of ubuntu:jammy? Once with the FROM (the 6 layers of the base image), and another one with the `COPY --from`, that includes the changes.

If the current final image works as expected, (replacing the whole root filesystem), then changing the base image for the final stage to `scratch`, ie: `FROM scratch`, that will work the same. 
`FROM scratch` means to base the image on nothing, your image won't have layers before it.

echoix commented on .config/gitpod/Dockerfile:
```diff
@@ -0,0 +1,100 @@
+
+COPY --from=gitpod-user / /
+
+USER gitpod
```
And an end of file without newline

echoix commented on .gitpod.yml:
When trying out on Gitpod, I observed that the installation of dependencies worked as expected. Then, the terminal that is kept open that executed the `make bootstrap` didn't have the python virtual environment available (even after pressing enter after an empty command). In fact, `which python` returned nothing.

![image](https://github.com/oxsecurity/megalinter/assets/27212526/75276fca-3b2d-4804-b20c-0b58c5bb9af6)

I had to open a new terminal (bash), that this one activates automatically without pressing enter:

![image](https://github.com/oxsecurity/megalinter/assets/27212526/ec8cc630-c006-4842-aa1f-e645766d68fb)
![image](https://github.com/oxsecurity/megalinter/assets/27212526/1a17d874-5653-4871-948d-876dd773ba45)
![image](https://github.com/oxsecurity/megalinter/assets/27212526/a879521e-71e4-495b-ad08-6e205e648bad)

echoix commented on .github/CONTRIBUTING.md:
```diff
@@ -3,49 +3,60 @@
-_(if you have a permission denied issue on Windows, please check [this solution](https://stackoverflow.com/a/57168165/7113625))_
+### 4. Test your changes
+
+#### 4.1 Gitpod
+
+Use Gitpod for a cloud-based development environment:
+
+1. Sign up for Gitpod: <https://gitpod.io>
+2. Fork the `megalinter` repository
+3. Open your fork in Gitpod: `https://gitpod.io/#https://github.com/username/megalinter`
+4. Create a new branch: `git checkout -b my-feature-branch`
+5. Make your changes and commit: `git add .` and `git commit -m "chore: description of changes"`
+6. Test all : `make tests` or `make tests-fast` for TDD mode
+7. Test with megalinter: `make megalinter-tests`
+8. Push your changes: `git push origin my-feature-branch`
+9. Create a pull request on GitHub
+10. Wait for a review
+
+Keep your Gitpod workspace synced with the main repository.
```
For trying out how Gitpod works for a repo you don't own, I think we can reformulate a little by not requiring that you sign up for Gitpod, but more like what we see in https://www.gitpod.io/docs/introduction/getting-started#step-1-your-first-workspace


> ## Step 1: Your first workspace
> 
> `youtube: ij1msCffQZA`
> 
> The best way to get see the power of Gitpod, is to try it out by starting your first [Workspace](/docs/configure/workspaces).
> 
> To start your first workspace:
> 
> 1. Navigate to a GitHub, GitLab or Bitbucket repository.
> 2. Open the repo in Gitpod by prefixing the URL with: `gitpod.io/#`.
> 3. Confirm the workspace creation.
> 
> For example, click this link to open the NodeJS project in a Gitpod workspace:
> 
> [`https://gitpod.io/#https://github.com/nodejs/node`](https://gitpod.io/#https://github.com/nodejs/node)
> 

And adding that you might need to press "Continue with GitHub" to log in, and authorize Gitpod (the permissions) it if it is the first time you use Gitpod.
![image](https://github.com/oxsecurity/megalinter/assets/27212526/40379827-b07e-43dc-8ef6-e29668a63b95)
![image](https://github.com/oxsecurity/megalinter/assets/27212526/b8430174-bc3e-4b0e-9390-eb144e5d8709)
and the last authorization page I can't see since I'm not new.

echoix commented:
So that's it, I made another round of review. I hope I wasn't too thorough ;)

nvuillam commented on .github/CONTRIBUTING.md:
I think a short video "how to contribute with gitpod" would be the best 😇

echoix commented on .config/make/gitpod.mak:
```diff
@@ -0,0 +1,13 @@
+## —— Gitpod —————————————————————————————————————————————————————————————————————————————————————
+.PHONY: gitpod-build
+gitpod-build: ## Run docker build image in local
+	docker build --tag postgresql_cluster:local --file .config/gitpod/Dockerfile .
```
Sorry, I wrote this comment 9days ago, but it wasn't like published..

echoix replied:
@ThomasSanson Is the tag the intended purpose?

echoix commented:
There's one piece I might want to understand a bit about your approach for testing. What does that new Gitpod workflow do to test? (More in a maintainer way, what does it do in order to know when to change something in it).

I think we don't need to keep building the image by ourselves at each commit (and publish it), since Gitpod will built it when needed. But that workflow doesn more than just build that Dockerfile.

echoix commented:
You can also place the PR in non-draft, so we could rerun CI with what we have now.

Is there anything else particularly that you'd like we review more carefully?

ThomasSanson commented:
@echoix 

For me, it is always important to ensure the viability of the contribution.

Assuming that we are working on an official contribution via Gitpod (but not limited to Gitpod, as we can contribute however we want), it is essential to test the proper functioning of the documentation and the ability to contribute via Gitpod.

That's why we are launching a GitHub action that will:
- Build the Gitpod image (without pushing it yet) to ensure that it still works.
- Lint the Dockerfile (by the way, it would make more sense to do this in reverse order, I'll fix that).

Indeed, it is not necessary all the time, hence my question, buried in multiple threads 😆, about the possibility of limiting this workflow only when the code changes in the .config folder.

ThomasSanson commented:
Good day,

I've been able to make headway on understanding Taskfile.

However, I suggest that we continue with the current pull request, and I will create another pull request for the update. This way, we're adhering to the KISS principle.

If this is acceptable to you, we can swiftly address the pending questions. If there are any similar to the one about migrating to Taskfile, we can open a new issue.

echoix commented:
> Good day,
> 
> 
> 
> I've been able to make headway on understanding Taskfile.
> 
> 
> 
> However, I suggest that we continue with the current pull request, and I will create another pull request for the update. This way, we're adhering to the KISS principle.
> 
> 
> 
> If this is acceptable to you, we can swiftly address the pending questions. If there are any similar to the one about migrating to Taskfile, we can open a new issue.

I'm comfortable with going with this intermediate step and continuing if needed.

echoix commented:
I think I have one last question about the *.mak files. Is there a reason that it is that extension (ie a convention I am not aware of yet)? Are they only plain makefiles? I didn't see that extension often and it seems nmake seem to use it (https://stackoverflow.com/questions/1438643/makefile-and-mak-file-codeblocks-and-vstudio).

I already saw .make and .mk (that one is used in the manual of GNU make https://www.gnu.org/software/make/manual/make.html#Include)

nvuillam commented:
@ThomasSanson if you merge main in your branch, CI should pass :)

ThomasSanson commented:
> I think I have one last question about the *.mak files. Is there a reason that it is that extension (ie a convention I am not aware of yet)? Are they only plain makefiles? I didn't see that extension often and it seems nmake seem to use it (https://stackoverflow.com/questions/1438643/makefile-and-mak-file-codeblocks-and-vstudio).
> 
> I already saw .make and .mk (that one is used in the manual of GNU make https://www.gnu.org/software/make/manual/make.html#Include)

The `.mak` suffix for makefiles isn't universally adopted, but finds application in certain scenarios. Its usage is contingent on the tools or environments employed.

Your inquiry suggests that `.mak` files are associated with the `nmake` utility, a Microsoft make tool. Typically, `nmake` employs makefiles with the `.mak` extension, though it isn't exclusive to it.

The linked StackOverflow thread delves into the use of `.mak` files in a cross-platform setting. The user aimed to create a project compatible with both Linux and Windows, utilizing `cbp2mak` to create a `.mak` file from a CodeBlocks project. This `.mak` file houses macros that decide the build, contingent on whether it's run on Windows or Linux, given platform-specific files.

Generally, the selection of makefile extension (.make, .mk, .mak, or even none) is primarily driven by the specific tool or environment norms. For instance, GNU `make` regularly employs files devoid of extension or with the `.mk` suffix as makefiles, but can accommodate any file type. Similarly, `nmake` frequently uses `.mak` files, but isn't confined to it.

To sum it up, `.mak` is merely one among the potential extensions for makefiles, employed in particular contexts such as with `nmake` or certain cross-platform situations. The makefile's functionality is not governed by its extension but its content, allowing different extensions to be swapped in most situations.

echoix commented:
I thinks everything is ready!

nvuillam commented:
IT's a little leap of faith , but as @ThomasSanson stays around for Makefile maintenance, and as the trusted @echoix approved... let's go ! :)
